src/core/util/latent_see.h

```diff
@@ -0,0 +1,213 @@
+};
+
+using ParentScope = Scope<true>;
+using InnerScope = Scope<false>;
+
+class Flow {
+ public:
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION Flow() : metadata_(nullptr) {}
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION explicit Flow(const Metadata* metadata)
+      : metadata_(metadata) {
+    Log::Append(metadata_, EventType::kFlowStart, id_);
+  }
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION ~Flow() {
+    if (metadata_ != nullptr) {
+      Log::Append(metadata_, EventType::kFlowEnd, id_);
+    }
+  }
+
+  Flow(const Flow&) = delete;
+  Flow& operator=(const Flow&) = delete;
+  Flow(Flow&& other) noexcept
+      : metadata_(std::exchange(other.metadata_, nullptr)), id_(other.id_) {}
+  Flow& operator=(Flow&& other) noexcept {
+    if (metadata_ != nullptr) Log::Append(metadata_, EventType::kFlowEnd, id_);
+    metadata_ = std::exchange(other.metadata_, nullptr);
+    id_ = other.id_;
+    return *this;
+  }
+
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION bool has_value() const {
+    return metadata_ != nullptr;
+  }
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION void reset() { emplace(nullptr); }
+  GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION void emplace(const Metadata* metadata) {
+    if (metadata_ != nullptr) Log::Append(metadata_, EventType::kFlowEnd, id_);
+    metadata_ = metadata;
+    if (metadata_ == nullptr) return;
+    id_ = next_flow_id_.fetch_add(1, std::memory_order_relaxed);
+    Log::Append(metadata_, EventType::kFlowStart, id_);
+  }
+
+ private:
+  const Metadata* metadata_;
+  uint64_t id_{next_flow_id_.fetch_add(1, std::memory_order_relaxed)};
+  static std::atomic<uint64_t> next_flow_id_;
+};
+
+GPR_ATTRIBUTE_ALWAYS_INLINE_FUNCTION inline void Mark(const Metadata* md) {
+  Log::Append(md, EventType::kMark, 0);
+}
+
+}  // namespace latent_see
+}  // namespace grpc_core
+#define GRPC_LATENT_SEE_METADATA(name)                                     \
+  []() {                                                                   \
+    static grpc_core::latent_see::Metadata metadata = {__FILE__, __LINE__, \
+                                                       #name};             \
+    return &metadata;                                                      \
+  }()
+// Parent scope: logs a begin and end event, and flushes the thread log on scope
+// exit. Because the flush takes some time it's better to place one parent scope
+// at the top of the stack, and use lighter weight scopes within it.
+#define GRPC_LATENT_SEE_PARENT_SCOPE(name)                       \
+  grpc_core::latent_see::ParentScope latent_see_scope##__LINE__( \
+      GRPC_LATENT_SEE_METADATA(name))
+// Inner scope: logs a begin and end event. Lighter weight than parent scope,
+// but does not flush the thread state - so should only be enclosed by a parent
+// scope.
+#define GRPC_LATENT_SEE_INNER_SCOPE(name)                       \
+  grpc_core::latent_see::InnerScope latent_see_scope##__LINE__( \
+      GRPC_LATENT_SEE_METADATA(name))
+// Mark: logs a single event.
```

Comment:
Nit: document that this is not automatically flushed.

Reply:
done.